From: Qi Zheng <zhengqi.arch@bytedance.com>
Subject: mm: vmscan: move shrinker_debugfs_remove() before synchronize_srcu()
Date: Thu, 1 Jun 2023 13:27:14 +0000

The debugfs_remove_recursive() will wait for debugfs_file_put() to return,
so there is no need to put it after synchronize_srcu() to wait for the rcu
read-side critical section to exit.

Just move it before synchronize_srcu(), which is also convenient to put
the heavy synchronize_srcu() in the delayed work later.

Link: https://lkml.kernel.org/r/20230601132714.2540707-1-qi.zheng@linux.dev
Signed-off-by: Qi Zheng <zhengqi.arch@bytedance.com>
Cc: Roman Gushchin <roman.gushchin@linux.dev>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/vmscan.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/mm/vmscan.c~mm-vmscan-move-shrinker_debugfs_remove-before-synchronize_srcu
+++ a/mm/vmscan.c
@@ -818,11 +818,11 @@ void unregister_shrinker(struct shrinker
 	debugfs_entry = shrinker_debugfs_detach(shrinker, &debugfs_id);
 	mutex_unlock(&shrinker_mutex);
 
+	shrinker_debugfs_remove(debugfs_entry, debugfs_id);
+
 	atomic_inc(&shrinker_srcu_generation);
 	synchronize_srcu(&shrinker_srcu);
 
-	shrinker_debugfs_remove(debugfs_entry, debugfs_id);
-
 	kfree(shrinker->nr_deferred);
 	shrinker->nr_deferred = NULL;
 }
_
